utils/validation.js

```diff
@@ -0,0 +1,261 @@
+function enforceConstraints(userId, planName, courses) {
+
+  return userConstraint(userId, planName, courses)
+    .then((conData) => {
+      return studentConstraint(conData[0], conData[1], conData[2]);
+    })
+    .then((conData) => {
+      return planNameConstraint(conData[0], conData[1], conData[2]);
+    })
+    .then((conData) => {
+      return zeroCourseConstraint(conData[0], conData[1], conData[2]);
+    })
+    .then((conData) => {
+      return duplicateCourseConstraint(conData[0], conData[1], conData[2]);
+    })
+    .then((conData) => {
+      return courseConstraint(conData[0], conData[1], conData[2]);
+    })
+    .then((conData) => {
+      return restrictionConstraint(conData[0], conData[1], conData[2]);
+    })
+    .then((conData) => {
+      return creditConstraint(conData[0], conData[1], conData[2]);
+    })
+    .then(() => {
+      console.log("Plan does not violate any constraints");
+      return 0;
+    })
+    .catch((conData) => {
+      // check if the error was from a constraint violation
+      if (conData[4]) {
+        return conData[4];
+      } else {
+        console.log("Error while trying to check constraints");
+        throw Error(conData[3]);
+      }
+
+    });
+
+}
+exports.enforceConstraints = enforceConstraints;
+
+// checks that the user exists
+function userConstraint(userId, planName, courses) {
+
+  return new Promise((resolve, reject) => {
+
+    const sql = "SELECT * FROM User WHERE userId=?;";
+    pool.query(sql, userId, (err, results) => {
+
+      if (err) {
+        console.log("Error checking user constraint");
+        reject([userId, planName, courses, err, 0]);
+      } else {
+
+        if (!results.length)
+          reject([userId, planName, courses, "", 1]);
+        else
+          resolve([userId, planName, courses, "", 0]);
+
+      }
+    });
+
+  });
+
+}
+
+// checks that the user is a student
+function studentConstraint(userId, planName, courses) {
+
+  return new Promise((resolve, reject) => {
+
+    const sql = "SELECT * FROM User WHERE userId=? AND role=0;";
+    pool.query(sql, userId, (err, results) => {
+
+      if (err) {
+        console.log("Error checking student constraint");
+        reject([userId, planName, courses, err, 0]);
+      } else {
+
+        if (!results.length)
+          reject([userId, planName, courses, "", 2]);
+        else
+          resolve([userId, planName, courses, "", 0]);
+
+      }
+    });
+
+  });
+
+}
+
+// checks that the plan name is a valid length
+function planNameConstraint(userId, planName, courses) {
+
+  return new Promise((resolve, reject) => {
+
+    if (planName.length < NAME_MIN || planName.length > NAME_MAX)
+      reject([userId, planName, courses, "", 3]);
+    else
+      resolve([userId, planName, courses, "", 0]);
+
+  });
+
+}
+
+// checks to see if any courses are selected
+function zeroCourseConstraint(userId, planName, courses) {
+
+  return new Promise((resolve, reject) => {
+
+    if (!courses.length)
```

Comment:
Should numerical checks be explicit? The idea of typecasting `null` and `undefined` to 0 here feels iffy. Maybe consider a default argument (e.g. `courses = []`)?